File: arrays/maximum-score-from-choosing-a-guarded-middle-segment-2/python/solution.py
"""
Title: Maximum Score from Choosing a Guarded Middle Segment

Problem Description:
You are given an integer array nums of length n and two non-negative integers L and R,
where 0 <= L, R < n. You must choose exactly one non-empty contiguous subarray nums[i..j]
as your active segment. The segment is considered valid only if there are at least L
elements strictly to its left and at least R elements strictly to its right. In other
words, the chosen segment must satisfy i >= L and j <= n - 1 - R.

The score of a valid segment is defined as:

    (minimum value inside the segment) * (sum of all values inside the segment)

Your task is to return the maximum possible score among all valid segments.

This is not simply a maximum-subarray problem, because the score depends on both the
segment sum and the segment minimum. Negative numbers are allowed, so extending a segment
may increase or decrease the score in non-obvious ways. An O(n^2) solution will not pass.

Constraints:
- 1 <= n <= 2 * 10^5
- -10^9 <= nums[i] <= 10^9
- 0 <= L, R < n
- There is guaranteed to be at least one valid segment

Notes:
- The answer may exceed 32-bit integer range, so use 64-bit arithmetic.
- A strong solution typically combines monotonic structure ideas with prefix sums and
  range optimization.
"""

from typing import List, Optional


class LiChaoNode:
    """
    Node for a Li Chao segment tree over a discrete set of x-coordinates.

    Each stored line has form:
        y = m * x + b

    The tree supports:
    - insert line
    - query maximum y at a given x

    This implementation is specialized for integer x-values taken from a fixed sorted list.
    """

    __slots__ = ("left", "right", "line")

    def __init__(self) -> None:
        """
        Initialize an empty Li Chao node.

        Args:
            None

        Returns:
            None

        Time complexity:
            O(1)

        Space complexity:
            O(1)
        """
        self.left: Optional["LiChaoNode"] = None
        self.right: Optional["LiChaoNode"] = None
        self.line: Optional[tuple[int, int]] = None


class LiChaoTree:
    """
    Li Chao tree for maximum queries on a fixed discrete x-domain.

    We use this to optimize expressions of the form:
        max over inserted lines (m * x + b)

    In this problem:
    - x will be a candidate minimum value nums[k]
    - lines are built from prefix sums
    """

    def __init__(self, xs: List[int]) -> None:
        """
        Create a Li Chao tree over the sorted unique x-values.

        Args:
            xs: Sorted unique x-coordinates where queries will happen.

        Returns:
            None

        Time complexity:
            O(1)

        Space complexity:
            O(1) excluding the input list reference
        """
        self.xs: List[int] = xs
        self.root: Optional[LiChaoNode] = None

    @staticmethod
    def _value(line: tuple[int, int], x: int) -> int:
        """
        Evaluate a line at x.

        Args:
            line: Tuple (m, b) representing y = m * x + b
            x: Query x-coordinate

        Returns:
            The value m * x + b

        Time complexity:
            O(1)

        Space complexity:
            O(1)
        """
        m, b = line
        return m * x + b

    def insert_line(self, m: int, b: int) -> None:
        """
        Insert a line y = m * x + b into the structure.

        Args:
            m: Slope
            b: Intercept

        Returns:
            None

        Time complexity:
            O(log M), where M is number of distinct x-values

        Space complexity:
            O(log M) recursion depth in worst case
        """
        line = (m, b)
        if self.root is None:
            self.root = LiChaoNode()
        self._insert(self.root, 0, len(self.xs) - 1, line)

    def _insert(self, node: LiChaoNode, left: int, right: int, line: tuple[int, int]) -> None:
        """
        Recursive helper for line insertion.

        Args:
            node: Current tree node
            left: Left index in xs
            right: Right index in xs
            line: Line to insert

        Returns:
            None

        Time complexity:
            O(log M)

        Space complexity:
            O(log M) recursion depth
        """
        if node.line is None:
            node.line = line
            return

        mid = (left + right) // 2
        x_left = self.xs[left]
        x_mid = self.xs[mid]
        x_right = self.xs[right]

        current = node.line

        # We maintain the line that is better at the midpoint in the current node.
        # If the new line is better at the midpoint, swap them.
        if self._value(line, x_mid) > self._value(current, x_mid):
            node.line, line = line, current
            current = node.line

        if left == right:
            return

        # Now decide where the remaining worse-at-mid line could still beat the stored line.
        # If it is better on the left endpoint, it may matter in the left child.
        if self._value(line, x_left) > self._value(node.line, x_left):
            if node.left is None:
                node.left = LiChaoNode()
            self._insert(node.left, left, mid, line)
        # Else if it is better on the right endpoint, it may matter in the right child.
        elif self._value(line, x_right) > self._value(node.line, x_right):
            if node.right is None:
                node.right = LiChaoNode()
            self._insert(node.right, mid + 1, right, line)

    def query(self, x: int) -> int:
        """
        Query the maximum y-value at x among all inserted lines.

        Args:
            x: Query x-coordinate

        Returns:
            Maximum value at x

        Time complexity:
            O(log M)

        Space complexity:
            O(1) iterative / O(log M) if counting recursion in helper
        """
        if self.root is None:
            raise ValueError("Query on empty Li Chao tree is not allowed.")

        # Binary search to find the discrete index of x.
        lo = 0
        hi = len(self.xs) - 1
        idx = -1
        while lo <= hi:
            mid = (lo + hi) // 2
            if self.xs[mid] == x:
                idx = mid
                break
            if self.xs[mid] < x:
                lo = mid + 1
            else:
                hi = mid - 1

        if idx == -1:
            raise ValueError("Query x must belong to the predefined coordinate set.")

        return self._query(self.root, 0, len(self.xs) - 1, idx, x)

    def _query(self, node: Optional[LiChaoNode], left: int, right: int, idx: int, x: int) -> int:
        """
        Recursive helper for maximum query.

        Args:
            node: Current tree node
            left: Left index in xs
            right: Right index in xs
            idx: Index of x in xs
            x: Actual x-coordinate

        Returns:
            Maximum line value at x from this subtree

        Time complexity:
            O(log M)

        Space complexity:
            O(log M) recursion depth
        """
        if node is None:
            return -(10 ** 40)

        best = -(10 ** 40)
        if node.line is not None:
            best = self._value(node.line, x)

        if left == right:
            return best

        mid = (left + right) // 2
        if idx <= mid:
            child_best = self._query(node.left, left, mid, idx, x)
        else:
            child_best = self._query(node.right, mid + 1, right, idx, x)

        return max(best, child_best)


class Solution:
    def maximumScore(self, nums: List[int], L: int, R: int) -> int:
        """
        Compute the maximum score of a valid guarded segment.

        The key idea is:
        1. For each index k, treat nums[k] as the minimum of the chosen segment.
        2. Using monotonic stack, find the maximal interval [left_bound, right_bound]
           where nums[k] is the minimum (with a careful tie-breaking rule so every
           segment is assigned to exactly one pivot index).
        3. Intersect that interval with the guard constraints:
               start >= L
               end   <= n - 1 - R
        4. Among all segments [i..j] inside that allowed interval and containing k,
           maximize:
               nums[k] * (prefix[j+1] - prefix[i])
        5. This becomes:
               if nums[k] >= 0:
                   nums[k] * (max prefix on right - min prefix on left)
               else:
                   nums[k] * (min prefix on right - max prefix on left)
        6. We answer these interval prefix min/max queries offline using Li Chao trees.

        Args:
            nums: Input integer array
            L: Minimum number of elements strictly to the left of chosen segment
            R: Minimum number of elements strictly to the right of chosen segment

        Returns:
            Maximum possible score as an integer

        Time complexity:
            O(n log n)

        Space complexity:
            O(n)
        """
        n = len(nums)

        # ------------------------------------------------------------
        # Step 1: Prefix sums
        #
        # prefix[t] = sum(nums[0:t])
        #
        # Then sum of subarray nums[i..j] is:
        #   prefix[j + 1] - prefix[i]
        #
        # We will repeatedly need subarray sums, so prefix sums are essential.
        # ------------------------------------------------------------
        prefix: List[int] = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]

        # ------------------------------------------------------------
        # Step 2: Compute ownership interval for each index as the chosen minimum.
        #
        # We need a standard "subarray minimum ownership" partition:
        # - prev strictly smaller element
        # - next smaller-or-equal element
        #
        # This tie-breaking ensures every subarray is assigned to exactly one index.
        #
        # For index k:
        #   left_bound  = prev_less[k] + 1
        #   right_bound = next_less_equal[k] - 1
        #
        # Any subarray [i..j] with:
        #   left_bound <= i <= k <= j <= right_bound
        # has nums[k] as the designated minimum owner.
        # ------------------------------------------------------------
        prev_less: List[int] = [-1] * n
        stack: List[int] = []
        for i in range(n):
            while stack and nums[stack[-1]] >= nums[i]:
                stack.pop()
            prev_less[i] = stack[-1] if stack else -1
            stack.append(i)

        next_less_equal: List[int] = [n] * n
        stack.clear()
        for i in range(n - 1, -1, -1):
            while stack and nums[stack[-1]] > nums[i]:
                stack.pop()
            next_less_equal[i] = stack[-1] if stack else n
            stack.append(i)

        # ------------------------------------------------------------
        # Step 3: Translate guard constraints.
        #
        # Valid segment [i..j] must satisfy:
        #   i >= L
        #   j <= n - 1 - R
        #
        # Let:
        #   max_end = n - 1 - R
        #
        # There is guaranteed to be at least one valid segment.
        # ------------------------------------------------------------
        max_end = n - 1 - R

        # ------------------------------------------------------------
        # Step 4: For each k, determine the allowed ranges of prefix indices.
        #
        # If a segment [i..j] is owned by k and valid, then:
        #   i in [max(left_bound, L), k]
        #   j in [k, min(right_bound, max_end)]
        #
        # In prefix-sum terms:
        #   left prefix index  = i
        #   right prefix index = j + 1
        #
        # So:
        #   i_prefix in [left_start, left_end] where
        #       left_start = max(left_bound, L)
        #       left_end   = k
        #
        #   r_prefix in [right_start, right_end] where
        #       right_start = k + 1
        #       right_end   = min(right_bound, max_end) + 1
        #
        # If either range is empty, index k cannot produce a valid segment.
        # ------------------------------------------------------------
        left_ranges: List[Optional[tuple[int, int]]] = [None] * n
        right_ranges: List[Optional[tuple[int, int]]] = [None] * n

        for k in range(n):
            left_bound = prev_less[k] + 1
            right_bound = next_less_equal[k] - 1

            left_start = max(left_bound, L)
            left_end = k

            right_start = k + 1
            right_end = min(right_bound, max_end) + 1

            if left_start <= left_end and right_start <= right_end:
                left_ranges[k] = (left_start, left_end)
                right_ranges[k] = (right_start, right_end)

        # ------------------------------------------------------------
        # Step 5: We need interval min/max queries on prefix sums, but the interval
        # depends on k and the final score multiplies by nums[k].
        #
        # For a fixed k:
        #   best_sum = best_right_prefix - best_left_prefix
        #
        # If nums[k] >= 0:
        #   maximize sum => maximize right prefix and minimize left prefix
        #
        # If nums[k] < 0:
        #   maximize nums[k] * sum => minimize sum
        #   so minimize (right prefix - left prefix)
        #   => minimize right prefix and maximize left prefix
        #
        # Therefore we need, for each k:
        #   max prefix on right interval
        #   min prefix on right interval
        #   max prefix on left interval
        #   min prefix on left interval
        #
        # We compute these offline using Li Chao trees.
        #
        # Why Li Chao?
        # Because each prefix value p at position t contributes:
        #   for max of nums[k] * p over active t, line y = p * x
        #   for min of nums[k] * p, we can maximize (-p) * x and negate later
        #
        # We process interval queries by sorting events by left endpoint and sweeping
        # the right endpoint, inserting active prefix positions into the tree.
        # ------------------------------------------------------------

        # All x-values where we will query are nums[k].
        xs = sorted(set(nums))

        # Prepare answer arrays for the four needed interval statistics.
        right_max_val: List[int] = [0] * n
        right_min_val: List[int] = [0] * n
        left_max_val: List[int] = [0] * n
        left_min_val: List[int] = [0] * n

        # ------------------------------------------------------------
        # Helper function:
        # Solve interval queries [ql, qr] on the prefix array values.
        #
        # For each query associated with index k and x = nums[k]:
        # - max_prefix_in_interval can be obtained by inserting lines y = prefix[t] * x
        #   and querying maximum at x = nums[k], then dividing conceptually by x is not
        #   needed because query directly returns prefix[t] * nums[k].
        #
        # But we actually need the raw prefix value, not multiplied.
        #
        # To avoid division/sign issues, we instead use:
        #   query max of p * x at x = nums[k]
        # and since x varies, this gives the correct "best weighted prefix" directly.
        #
        # For the final score formulas:
        #   if nums[k] >= 0:
        #       score = max_right_weighted - min_left_weighted
        #   if nums[k] < 0:
        #       score = min_right_weighted - max_left_weighted
        #
        # where:
        #   max_right_weighted = nums[k] * max(prefix on right)
        #   min_right_weighted = nums[k] * min(prefix on right)
        # etc.
        #
        # This is exactly what Li Chao gives us efficiently.
        # ------------------------------------------------------------
        def solve_weighted_interval_queries(
            ranges: List[Optional[tuple[int, int]]],
            want_max_weighted: bool,
        ) -> List[int]:
            """
            Offline interval query solver for weighted prefix optimization.

            For each index k with interval [l, r], compute:
                if want_max_weighted:
                    max over t in [l, r] of nums[k] * prefix[t]
                else:
                    min over t in [l, r] of nums[k] * prefix[t]

            Args:
                ranges: List of optional intervals on prefix indices
                want_max_weighted: Whether to compute weighted maximum or minimum

            Returns:
                List ans where ans[k] is the requested weighted optimum

            Time complexity:
                O(n log n)

            Space complexity:
                O(n)
            """
            queries_by_left: List[List[tuple[int, int]]] = [[] for _ in range(n + 1)]
            for k, interval in enumerate(ranges):
                if interval is not None:
                    l, r = interval
                    queries_by_left[l].append((r, k))

            # We sweep left boundary from n down to 0.
            #
            # At sweep position current_left, the active set contains all prefix indices t
            # such that t >= current_left and already inserted.
            #
            # To answer interval [l, r], we need only t in [l, r].
            # So we group queries by l and use a Fenwick-like idea? Not enough for weighted
            # optimization with varying x.
            #
            # Instead, we use a divide-and-conquer style sweep with right endpoint ordering:
            # Here we choose a simpler and still O(n log n) method:
            # process queries sorted by r, while inserting positions in increasing order,
            # but to enforce lower bound l we run a second pass on reversed indices.
            #
            # However

File: arrays/maximum-score-from-choosing-a-guarded-middle-segment-2/java/Solution.java
import java.util.*;

/*
Problem Title: Maximum Score from Choosing a Guarded Middle Segment

Problem Description:
You are given an integer array nums of length n and two non-negative integers L and R, where 0 <= L, R < n.
You must choose exactly one non-empty contiguous subarray nums[i..j] as your active segment.
The segment is considered valid only if there are at least L elements strictly to its left and at least R elements strictly to its right.
In other words, the chosen segment must satisfy i >= L and j <= n - 1 - R.

The score of a valid segment is defined as:

(minimum value inside the segment) * (sum of all values inside the segment)

Your task is to return the maximum possible score among all valid segments.

This is not simply a maximum-subarray problem, because the score depends on both the segment sum and the segment minimum.
Negative numbers are allowed, so extending a segment may increase or decrease the score in non-obvious ways.
An O(n^2) solution will not pass.

Constraints:
- 1 <= n <= 2 * 10^5
- -10^9 <= nums[i] <= 10^9
- 0 <= L, R < n
- There is guaranteed to be at least one valid segment

Notes:
- The answer may exceed 32-bit integer range, so use 64-bit arithmetic.
- A strong solution typically combines monotonic structure ideas with prefix sums and range optimization.

Core idea of this solution:
1) For every index k, treat nums[k] as the minimum of the chosen segment.
2) Using monotonic stacks, compute the maximal interval [leftBound[k], rightBound[k]] in which nums[k]
   is the unique chosen representative minimum:
      - all values inside that interval are >= nums[k]
      - ties are broken consistently using:
            previous strictly smaller
            next smaller-or-equal
3) Any valid segment where nums[k] is the minimum must satisfy:
      left in [max(L, leftBound[k]), min(k, n-1-R)]
      right in [max(k, L), min(rightBound[k], n-1-R)]
   and left <= k <= right.
4) Let prefix sums be P, so sum(left..right) = P[right+1] - P[left].
   For fixed k:
      score = nums[k] * (P[right+1] - P[left])
   Therefore:
      - if nums[k] >= 0, maximize the segment sum
        => choose maximum P[right+1] and minimum P[left]
      - if nums[k] < 0, minimize the segment sum
        => choose minimum P[right+1] and maximum P[left]
5) So for each k we only need range min/max queries on prefix sums.
   We build segment trees over prefix sums to answer:
      min / max on P indices corresponding to allowed left endpoints
      min / max on P indices corresponding to allowed right+1 endpoints
6) This yields O(n log n) time.

Important note about Example 1 in the prompt:
The textual explanation is inconsistent with the formal rule.
This implementation follows the formal validity rule exactly, as required.
*/

public class Solution {

    /**
     * Solves the problem: maximum score among all valid guarded subarrays.
     *
     * Detailed meaning:
     * - We may only choose subarrays nums[i..j] such that i >= L and j <= n - 1 - R.
     * - Score = (minimum element in subarray) * (sum of subarray).
     *
     * Algorithm summary:
     * 1. Compute prefix sums.
     * 2. Compute, for each index k, the maximal interval where nums[k] can serve as the chosen minimum
     *    under a consistent tie-breaking rule.
     * 3. For each k, derive the allowed range of left endpoints and right endpoints.
     * 4. Depending on the sign of nums[k], either maximize or minimize the subarray sum.
     * 5. Use segment trees on prefix sums to query range min/max efficiently.
     *
     * @param nums the input array
     * @param L minimum number of elements strictly to the left of the chosen segment
     * @param R minimum number of elements strictly to the right of the chosen segment
     * @return the maximum possible score using 64-bit arithmetic
     * Time complexity: O(n log n)
     * Space complexity: O(n)
     */
    public long maximumScore(int[] nums, int L, int R) {
        int n = nums.length;

        // Prefix sums:
        // prefix[t] = sum of nums[0..t-1]
        // Therefore sum of nums[i..j] = prefix[j+1] - prefix[i]
        long[] prefix = buildPrefixSums(nums);

        // For each index i:
        // leftLess[i]  = index of previous element strictly smaller than nums[i], or -1 if none
        // rightLessEq[i] = index of next element smaller than or equal to nums[i], or n if none
        //
        // Then nums[i] is the designated minimum for any subarray [l..r] such that:
        // leftLess[i] < l <= i <= r < rightLessEq[i]
        int[] leftLess = previousStrictlySmaller(nums);
        int[] rightLessEq = nextSmallerOrEqual(nums);

        // Convert the above strict/open boundaries into inclusive allowed bounds for endpoints.
        int[] leftBound = new int[n];
        int[] rightBound = new int[n];
        for (int i = 0; i < n; i++) {
            leftBound[i] = leftLess[i] + 1;
            rightBound[i] = rightLessEq[i] - 1;
        }

        // Build segment trees over prefix sums.
        //
        // Why prefix sums?
        // For a subarray [l..r], sum = prefix[r+1] - prefix[l].
        //
        // So for each i we need:
        // - queries over prefix[l] where l is in some interval
        // - queries over prefix[r+1] where r is in some interval, meaning (r+1) is also an interval
        SegmentTreeMin segMin = new SegmentTreeMin(prefix);
        SegmentTreeMax segMax = new SegmentTreeMax(prefix);

        long answer = Long.MIN_VALUE;

        // Valid segment must satisfy:
        // start >= L
        // end   <= n - 1 - R
        int globalMinStart = L;
        int globalMaxEnd = n - 1 - R;

        // Guaranteed at least one valid segment exists by the problem statement.
        // We now evaluate every index as the designated minimum.
        for (int i = 0; i < n; i++) {
            // Allowed left endpoints for segments where nums[i] is the designated minimum:
            // left must be:
            //   - within the interval where nums[i] remains minimum: [leftBound[i], i]
            //   - valid under guard rule: left >= L
            int leftStart = Math.max(leftBound[i], globalMinStart);
            int leftEnd = i;

            // Allowed right endpoints:
            //   - within the interval where nums[i] remains minimum: [i, rightBound[i]]
            //   - valid under guard rule: right <= n - 1 - R
            int rightStart = i;
            int rightEnd = Math.min(rightBound[i], globalMaxEnd);

            // If either side has no feasible endpoint, then no valid segment uses i as designated minimum.
            if (leftStart > leftEnd || rightStart > rightEnd) {
                continue;
            }

            // We need prefix[left] where left in [leftStart, leftEnd]
            int prefixLeftL = leftStart;
            int prefixLeftR = leftEnd;

            // We need prefix[right+1] where right in [rightStart, rightEnd]
            // Therefore prefix index is in [rightStart + 1, rightEnd + 1]
            int prefixRightL = rightStart + 1;
            int prefixRightR = rightEnd + 1;

            long minLeftPrefix = segMin.query(prefixLeftL, prefixLeftR);
            long maxLeftPrefix = segMax.query(prefixLeftL, prefixLeftR);
            long minRightPrefix = segMin.query(prefixRightL, prefixRightR);
            long maxRightPrefix = segMax.query(prefixRightL, prefixRightR);

            long value = nums[i];
            long bestSumForThisMinimum;

            // score = value * (prefix[right+1] - prefix[left])
            //
            // If value >= 0:
            //   maximize sum => maximize prefix[right+1] and minimize prefix[left]
            //
            // If value < 0:
            //   maximize value * sum => since value is negative, minimize sum
            //   => minimize prefix[right+1] and maximize prefix[left]
            if (value >= 0) {
                bestSumForThisMinimum = maxRightPrefix - minLeftPrefix;
            } else {
                bestSumForThisMinimum = minRightPrefix - maxLeftPrefix;
            }

            long score = value * bestSumForThisMinimum;
            answer = Math.max(answer, score);
        }

        return answer;
    }

    /**
     * Builds prefix sums for the array.
     *
     * prefix[0] = 0
     * prefix[i+1] = nums[0] + nums[1] + ... + nums[i]
     *
     * @param nums the input array
     * @return prefix sum array of length nums.length + 1
     * Time complexity: O(n)
     * Space complexity: O(n)
     */
    public long[] buildPrefixSums(int[] nums) {
        int n = nums.length;
        long[] prefix = new long[n + 1];
        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }
        return prefix;
    }

    /**
     * Computes the previous index with a strictly smaller value for every position.
     *
     * Example:
     * nums = [5, 2, 4, 3]
     * result might be [-1, -1, 1, 1]
     *
     * We use a monotonic increasing stack.
     * While stack top >= current value, pop.
     * Then the remaining top, if any, is strictly smaller.
     *
     * @param nums the input array
     * @return array prev where prev[i] is the previous index j < i with nums[j] < nums[i], or -1
     * Time complexity: O(n)
     * Space complexity: O(n)
     */
    public int[] previousStrictlySmaller(int[] nums) {
        int n = nums.length;
        int[] prev = new int[n];
        Arrays.fill(prev, -1);

        Deque<Integer> stack = new ArrayDeque<>();

        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] >= nums[i]) {
                stack.pop();
            }
            prev[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(i);
        }

        return prev;
    }

    /**
     * Computes the next index with a smaller-or-equal value for every position.
     *
     * This tie-breaking pairs with previousStrictlySmaller() so that equal values are assigned
     * consistently to exactly one representative minimum index.
     *
     * We scan from right to left with a monotonic increasing stack.
     * While stack top > current value, pop.
     * Then the remaining top, if any, is <= current value.
     *
     * @param nums the input array
     * @return array next where next[i] is the next index j > i with nums[j] <= nums[i], or n
     * Time complexity: O(n)
     * Space complexity: O(n)
     */
    public int[] nextSmallerOrEqual(int[] nums) {
        int n = nums.length;
        int[] next = new int[n];
        Arrays.fill(next, n);

        Deque<Integer> stack = new ArrayDeque<>();

        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && nums[stack.peek()] > nums[i]) {
                stack.pop();
            }
            next[i] = stack.isEmpty() ? n : stack.peek();
            stack.push(i);
        }

        return next;
    }

    /**
     * Brute-force verifier for small arrays.
     * This is useful for demonstration and sanity checking.
     *
     * It checks every valid subarray, computes its minimum and sum directly,
     * and returns the best score.
     *
     * @param nums the input array
     * @param L minimum number of elements strictly to the left
     * @param R minimum number of elements strictly to the right
     * @return the exact answer by brute force
     * Time complexity: O(n^3) in the worst case
     * Space complexity: O(1) extra space
     */
    public long maximumScoreBruteForce(int[] nums, int L, int R) {
        int n = nums.length;
        long ans = Long.MIN_VALUE;

        for (int i = L; i < n; i++) {
            for (int j = i; j <= n - 1 - R; j++) {
                long sum = 0;
                int min = Integer.MAX_VALUE;
                for (int k = i; k <= j; k++) {
                    sum += nums[k];
                    min = Math.min(min, nums[k]);
                }
                ans = Math.max(ans, sum * (long) min);
            }
        }

        return ans;
    }

    /**
     * Demonstrates the solution on sample-style inputs and prints the results.
     *
     * Note:
     * The first example text in the prompt is internally inconsistent.
     * This main method prints the result according to the formal rule only.
     *
     * @param args command-line arguments (unused)
     * @return nothing
     * Time complexity: O(1) for the fixed demonstrations, excluding the solver calls
     * Space complexity: O(1) extra space, excluding the solver internals
     */
    public static void main(String[] args) {
        Solution sol = new Solution();

        int[] nums1 = {5, 2, 4, 3, 6};
        int L1 = 1, R1 = 1;
        long ans1 = sol.maximumScore(nums1, L1, R1);
        long brute1 = sol.maximumScoreBruteForce(nums1, L1, R1);
        System.out.println("Example 1 (formal rule) result: " + ans1);
        System.out.println("Example 1 brute-force check: " + brute1);

        int[] nums2 = {4, -1, 2, -2, 5};
        int L2 = 0, R2 = 0;
        long ans2 = sol.maximumScore(nums2, L2, R2);
        long brute2 = sol.maximumScoreBruteForce(nums2, L2, R2);
        System.out.println("Example 2 result: " + ans2);
        System.out.println("Example 2 brute-force check: " + brute2);

        int[] nums3 = {1};
        int L3 = 0, R3 = 0;
        System.out.println("Single element result: " + sol.maximumScore(nums3, L3, R3));

        int[] nums4 = {-3, -2, -5, -1};
        int L4 = 0, R4 = 0;
        System.out.println("All negative result: " + sol.maximumScore(nums4, L4, R4));
        System.out.println("All negative brute-force: " + sol.maximumScoreBruteForce(nums4, L4, R4));
    }

    /**
     * Segment tree supporting range minimum queries on a long array.
     */
    static class SegmentTreeMin {
        private final int n;
        private final long[] tree;

        /**
         * Builds a segment tree for range minimum queries.
         *
         * @param arr source array
         * @return nothing
         * Time complexity: O(n)
         * Space complexity: O(n)
         */
        SegmentTreeMin(long[] arr) {
            int size = 1;
            while (size < arr.length) {
                size <<= 1;
            }
            this.n = size;
            this.tree = new long[n << 1];
            Arrays.fill(tree, Long.MAX_VALUE);

            for (int i = 0; i < arr.length; i++) {
                tree[n + i] = arr[i];
            }
            for (int i = n - 1; i >= 1; i--) {
                tree[i] = Math.min(tree[i << 1], tree[i << 1 | 1]);
            }
        }

        /**
         * Returns the minimum value in arr[left..right], inclusive.
         *
         * @param left left index, inclusive
         * @param right right index, inclusive
         * @return minimum value in the range
         * Time complexity: O(log n)
         * Space complexity: O(1)
         */
        long query(int left, int right) {
            long res = Long.MAX_VALUE;
            int l = left + n;
            int r = right + n;

            while (l <= r) {
                if ((l & 1) == 1) {
                    res = Math.min(res, tree[l++]);
                }
                if ((r & 1) == 0) {
                    res = Math.min(res, tree[r--]);
                }
                l >>= 1;
                r >>= 1;
            }

            return res;
        }
    }

    /**
     * Segment tree supporting range maximum queries on a long array.
     */
    static class SegmentTreeMax {
        private final int n;
        private final long[] tree;

        /**
         * Builds a segment tree for range maximum queries.
         *
         * @param arr source array
         * @return nothing
         * Time complexity: O(n)
         * Space complexity: O(n)
         */
        SegmentTreeMax(long[] arr) {
            int size = 1;
            while (size < arr.length) {
                size <<= 1;
            }
            this.n = size;
            this.tree = new long[n << 1];
            Arrays.fill(tree, Long.MIN_VALUE);

            for (int i = 0; i < arr.length; i++) {
                tree[n + i] = arr[i];
            }
            for (int i = n - 1; i >= 1; i--) {
                tree[i] = Math.max(tree[i << 1], tree[i << 1 | 1]);
            }
        }

        /**
         * Returns the maximum value in arr[left..right], inclusive.
         *
         * @param left left index, inclusive
         * @param right right index, inclusive
         * @return maximum value in the range
         * Time complexity: O(log n)
         * Space complexity: O(1)
         */
        long query(int left, int right) {
            long res =